### PR TITLE
new resource ons instance 

### PR DESCRIPTION
**Pull request by supercym**
add ons-client support
add resource ons-instance and its test/import test/service/example/website files
Acceptance Test:
Region in Beijing(cn-beijing)
image
Region in London(eu-west-1)
image

### CODE REVIEW

File: alicloud/resource_alicloud_ons_instance.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	request.PreventCache = onsService.GetPreventCache()
+	if v, ok := d.GetOk("remark"); ok {
+		request.Remark = v.(string)
+	}
+
+	raw, err := client.WithOnsClient(func(onsClient *ons.Client) (interface{}, error) {
+		return onsClient.OnsInstanceCreate(request)
+	})
+	if err != nil {
+		return fmt.Errorf("create ons instance got an error: %#v", err)
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ons_instance.go]: pleasw use WrapErrorf()

File: alicloud/resource_alicloud_ons_instance.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	}
+
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ons_instance.go]: the line 76,add addDebug()

File: alicloud/resource_alicloud_ons_instance.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+	})
+
+	if err != nil {
+		// Handle exceptions
+		instanceNotExists := onsService.InstanceNotExistFunc(err)
+		if instanceNotExists {
+			d.SetId("")
+			return nil
+		} else {
+			return fmt.Errorf("get ons instance base info error: %#v", err)
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ons_instance.go]: please use WrapError()

File: alicloud/resource_alicloud_ons_instance.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+	request.PreventCache = onsService.GetPreventCache()
+
+	_, err := client.WithOnsClient(func(onsClient *ons.Client) (interface{}, error) {
+		return onsClient.OnsInstanceDelete(request)
+	})
+	if err != nil {
+		return fmt.Errorf("delete instance error: %#v", err)
+	}
+
+	return resourceAlicloudONSInstanceRead(d, meta)
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ons_instance.go]: this not need read,but should to check successful of delete by watifor()

File: alicloud/resource_alicloud_ons_instance.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+	request.InstanceId = d.Get("instance_id").(string)
+	request.PreventCache = onsService.GetPreventCache()
+
+	_, err := client.WithOnsClient(func(onsClient *ons.Client) (interface{}, error) {
+		return onsClient.OnsInstanceDelete(request)
+	})
+	if err != nil {
+		return fmt.Errorf("delete instance error: %#v", err)
+	}
+
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ons_instance.go]: stil,line 180 use WrapErrorf(),and line 182 add addDebug()

File: alicloud/import_alicloud_ons_instance_test.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+		CheckDestroy: testAccCheckOnsInstanceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccOnsInstanceConfig(acctest.RandIntRange(10000, 999999)),
+			},
+
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
```
Comment (YY-tomorrow) [inline alicloud/import_alicloud_ons_instance_test.go]: With the new format, no need import, reference dns_records

File: alicloud/resource_alicloud_ons_instance.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+	if attributeUpdate {
+		_, err := client.WithOnsClient(func(onsClient *ons.Client) (interface{}, error) {
+			return onsClient.OnsInstanceUpdate(request)
+		})
+		if err != nil {
+			return fmt.Errorf("update instance error: %#v", err)
+		}
+	}
+
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ons_instance.go]: still

File: alicloud/resource_alicloud_ons_instance.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+	d.Set("release_time", time.Unix(int64(response.InstanceBaseInfo.ReleaseTime)/1000, 0).Format("2006-01-02 03:04:05"))
+
+	return nil
+}
+
+func resourceAlicloudONSInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	onsService := OnsService{}
+
+	d.Partial(true)
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ons_instance.go]: i think partial is no need

File: alicloud/resource_alicloud_ons_instance.go [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+	attributeUpdate := false
+
+	request := ons.CreateOnsInstanceUpdateRequest()
+	request.InstanceId = d.Get("instance_id").(string)
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ons_instance.go]: if instance_id no support update,please set be forcenew

File: alicloud/resource_alicloud_ons_instance_test.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+		if err != nil {
+			log.Printf("[ERROR] Failed to delete ons instance (%s): %s", name, err)
+		}
+	}
+
+	return nil
+}
+
+func TestAccAlicloudOnsInstance_basic(t *testing.T) {
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ons_instance_test.go]: please reference dns_records, modify this testcase

File: website/docs/r/ons_instance.html.markdown [hunk 11]
```diff
@@ -0,0 +1,48 @@
+---
+layout: "alicloud"
+page_title: "Alicloud: alicloud_ons_instance"
+sidebar_current: "docs-alicloud-resource-ons-instance"
+description: |-
+  Provides a Alicloud ONS Instance resource.
+---
+
```
Comment (YY-tomorrow) [inline website/docs/r/ons_instance.html.markdown]: please add a note, avaliable in 1.49.0+

File: alicloud/resource_alicloud_ons_instance.go [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+	onsService := OnsService{client}
+
+	response, err := onsService.DescribeOnsInstance(d.Id())
+
+	if err != nil {
+		// Handle exceptions
+		if IsExceptedErrors(err, []string{OnsInstanceNotExist}) {
+			return nil
+		} else {
+			return WrapErrorf(err, "get ons instance base info error.")
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ons_instance.go]: use WrapError()

File: alicloud/resource_alicloud_ons_instance_test.go [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+
+		if err != nil {
+			log.Printf("[ERROR] Failed to delete ons instance (%s): %s", name, err)
+		}
+	}
+
+	return nil
+}
+
+var randNum int
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_ons_instance_test.go]: please change randNum to rand

File: alicloud/service_alicloud_ons.go [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+func (s *OnsService) InstanceNotExistFunc(err error) bool {
+	return strings.Contains(err.Error(), OnsInstanceNotExist)
+}
+
+func (s *OnsService) GetPreventCache() requests.Integer {
+	return requests.NewInteger(int(time.Now().UnixNano() / 1e6))
+}
+
+func (s *OnsService) DescribeOnsInstance(instance_id string) (*ons.OnsInstanceBaseInfoResponse, error) {
```
Comment (YY-tomorrow) [inline alicloud/service_alicloud_ons.go]: change instance_id to id

File: alicloud/resource_alicloud_ons_instance.go [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+				Optional:     true,
+				ValidateFunc: validateOnsInstanceRemark,
+			},
+
+			// Computed Values
+			"instance_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ons_instance.go]: `instance_id` is needless and `id` will replace it.

File: alicloud/resource_alicloud_ons_instance.go [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+import (
+	"fmt"
+	"time"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ons"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-alicloud/alicloud/connectivity"
+)
+
+func resourceAlicloudONSInstance() *schema.Resource {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ons_instance.go]: Please use `resourceAlicloudOnsInstance`

File: alicloud/resource_alicloud_ons_instance.go [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+func resourceAlicloudONSInstanceRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	onsService := OnsService{client}
+
+	response, err := onsService.DescribeOnsInstance(d.Id())
+
+	if err != nil {
+		// Handle exceptions
+		if IsExceptedErrors(err, []string{OnsInstanceNotExist}) {
+			return nil
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ons_instance.go]: I think this shoud be 
```
if NotFoundError(err) {
   d.SetId("")
   return nil
}
return WrapError(err)
```

File: alicloud/resource_alicloud_ons_instance.go [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+func resourceAlicloudONSInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	onsService := OnsService{client}
+
+	attributeUpdate := false
+
+	request := ons.CreateOnsInstanceUpdateRequest()
+	request.InstanceId = d.Get("instance_id").(string)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ons_instance.go]: I think the line 116 should be `request.InstanceId = d.Id()`

File: alicloud/resource_alicloud_ons_instance.go [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+	request.PreventCache = onsService.GetPreventCache()
+
+	if d.HasChange("name") {
+		d.SetPartial("name")
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ons_instance.go]: Please remove `d.SetPartial`

File: alicloud/resource_alicloud_ons_instance.go [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+		var name string
+		if v, ok := d.GetOk("name"); ok {
+			name = v.(string)
+		} else {
+			return fmt.Errorf("cann't change name to empty string")
+		}
+		request.InstanceName = name
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ons_instance.go]: The line 124-125 are redundant. Please remove them.

File: alicloud/resource_alicloud_ons_instance.go [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+	return
+}
+
+func validateOnsInstanceRemark(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+	if len(value) > 128 {
+		errors = append(errors, fmt.Errorf("%q cannot be longer than 128 characters", k))
+	}
+	return
+}
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ons_instance.go]: Please move the func `validateOnsInstanceName ` and `validateOnsInstanceRemark ` to validator.go

File: alicloud/resource_alicloud_ons_instance_test.go [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+	}
+
+	request := ons.CreateOnsInstanceInServiceListRequest()
+	request.PreventCache = onsService.GetPreventCache()
+
+	raw, err := onsService.client.WithOnsClient(func(onsClient *ons.Client) (interface{}, error) {
+		return onsClient.OnsInstanceInServiceList(request)
+	})
+	if err != nil {
+		return WrapErrorf(err, "error retrieving ONS instance in service list.")
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ons_instance_test.go]: Don't return this error and just print it is ok.

File: website/docs/r/ons_instance.html.markdown [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+# alicloud\_ons\_instance
+
+Provides a ONS instance resource.
+
+-> **NOTE:**  The number of instances in the same region cannot exceed 8. Does not support region "mq-internet-access" and "ap-southeast-5".
```
Comment (xiaozhu36) [inline website/docs/r/ons_instance.html.markdown]: The number of instances in the same region cannot exceed 8. At present, the resource does not support region "mq-internet-access" and "ap-southeast-5".

File: alicloud/provider.go [hunk 24]
```diff
@@ -667,6 +671,12 @@ func endpointsSchema() *schema.Schema {
 					Default:     "",
 					Description: descriptions["oss_endpoint"],
 				},
+				"ons": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					Default:     "",
+					Description: descriptions["ons_endpoint"],
+				},
```
Comment (xiaozhu36) [inline alicloud/provider.go]: missing description "ons_endpoint"

File: alicloud/resource_alicloud_ons_instance.go [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+		request.InstanceName = name
+		attributeUpdate = true
+	}
+
+	if d.HasChange("remark") {
+		var remark string
+		if v, ok := d.GetOk("remark"); ok {
+			remark = v.(string)
+		} else {
+			return fmt.Errorf("cann't change remark to empty string")
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ons_instance.go]: still applies

File: website/docs/r/ons_instance.html.markdown [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+layout: "alicloud"
+page_title: "Alicloud: alicloud_ons_instance"
+sidebar_current: "docs-alicloud-resource-ons-instance"
+description: |-
+  Provides a Alicloud ONS Instance resource.
+---
+
+# alicloud\_ons\_instance
+
+Provides a ONS instance resource.
```
Comment (xiaozhu36) [inline website/docs/r/ons_instance.html.markdown]: Please add a document link for this resource.

File: alicloud/resource_alicloud_ons_instance.go [hunk 27]
```diff
@@ -0,0 +1,172 @@
...
+
+	raw, err := onsService.client.WithOnsClient(func(onsClient *ons.Client) (interface{}, error) {
+		return onsClient.OnsInstanceDelete(request)
+	})
+	if err != nil {
+		if IsExceptedErrors(err, []string{OnsInstanceNotExist}) {
+			return nil
+		}
+		if IsExceptedErrors(err, []string{OnsInstanceNotEmpty}) {
+			_, err := onsService.DeleteOnsInstance(d.Id())
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ons_instance.go]: There is no need to delete topic in the instance resource. It should be deleted in the topic resource and there should add retry for the error OnsInstanceNotEmpty, like vpc and vswitch.

File: website/alicloud.erb [hunk 28]
```diff
@@ -662,6 +662,15 @@
                     </ul>
                 </li>
 
+                <li<%= sidebar_current("docs-alicloud-resource-ons") %>>
+                    <a href="#">ONS Resources</a>
```
Comment (xiaozhu36) [inline website/alicloud.erb]: Please modify line 666 to ` <a href="#">RocketMQ Resources</a>`

File: website/alicloud.erb [hunk 29]
```diff
@@ -662,6 +662,15 @@
...
+                    <ul class="nav nav-visible">
```
Comment (xiaozhu36) [inline website/alicloud.erb]: Please remove `nav-visible`

**Pull request merged**